### PR TITLE
[plugin.video.youtube] v6.1.2

### PR DESCRIPTION
**Pull request by anxdpanic**
### Description
Changelog: https://github.com/jdf76/plugin.video.youtube/blob/6.x.x/changelog.txt#L1-L38

### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

**Comment by anxdpanic** [comment]
I've fixed the complex entry warning and the codacy issue, leaving the commit separate in case anyone has started reviewing. Will bump version and squash when given the go ahead.

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py [hunk 1]
```diff
@@ -7,6 +7,75 @@
...
+    list_item.setProperty(u'IsPlayable', u'true')
+
+    if play_item.get_fanart() and settings.show_fanart():
+        fanart = play_item.get_fanart()
+    if major_version <= 12:
+        list_item.setIconImage(thumb)
+        list_item.setProperty("Fanart_Image", fanart)
+    elif major_version <= 15:
+        list_item.setArt({'thumb': thumb, 'fanart': fanart})
+        list_item.setIconImage(thumb)
```
Comment (enen92) [inline plugin.video.youtube/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py]: replace by `icon`:`xxx` above

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py [hunk 2]
```diff
@@ -7,6 +7,75 @@
...
+    else:
+        list_item.setArt({'icon': thumb, 'thumb': thumb, 'fanart': fanart})
+
+    if not play_item.use_dash() and not settings.is_support_alternative_player_enabled() and \
+            play_item.get_headers() and play_item.get_uri().startswith('http'):
+        play_item.set_uri(play_item.get_uri() + '|' + play_item.get_headers())
+
+    list_item.setProperty('inputstreamaddon', '')
```
Comment (enen92) [inline plugin.video.youtube/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py]: is this needed?

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/service.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    context.log_debug('YouTube service initialization...')
+    version = context.get_system_version().get_version()
+    settings = context.get_settings()
+
+    player = YouTubePlayer(context=context)
+    monitor = YouTubeMonitor()
+
+    monitor.remove_temp_dir()
+
+    while not monitor.abortRequested():
```
Comment (enen92) [inline plugin.video.youtube/resources/lib/youtube_plugin/kodion/service.py]: I don't think this works across all versions but is fine according to the branch you are submitting. Probably the checks you have above for the listitem art stuff for legacy versions of kodi can be removed

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/utils/function_cache.py [hunk 4]
```diff
@@ -106,16 +105,4 @@ def _seconds_difference(_first, _last):
         return cached_data
 
     def _optimize_item_count(self):
-        clear = False
```
Comment (enen92) [inline plugin.video.youtube/resources/lib/youtube_plugin/kodion/utils/function_cache.py]: probably you don't need the method

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/utils/ipstack.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
+import requests
+
+from base64 import b64decode
+
+
+__API_KEY = 'ZGFiODVhMzM5NTI2ZmVhNmEwZDc5ZjkzMjM2M2FjY2Q='
+__API_BASE_URL = 'http://api.ipstack.com'
```
Comment (enen92) [inline plugin.video.youtube/resources/lib/youtube_plugin/kodion/utils/ipstack.py]: What is this for?

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/utils/player.py [hunk 6]
```diff
@@ -9,25 +9,82 @@ def __init__(self, *args, **kwargs):
...
+        self.current_video_total_time = self.getTotalTime()
+        while self.isPlaying():
+            xbmc.sleep(500)
+            if self.isPlaying():
+                self.current_played_time = self.getTime()
+                if self.current_video_total_time == 0.0:
+                    self.current_video_total_time = self.getTotalTime()
+
+    def onAVStarted(self):
+        if self.context.get_settings().use_playback_history():
```
Comment (enen92) [inline plugin.video.youtube/resources/lib/youtube_plugin/kodion/utils/player.py]: only v18+, please restrict if possible

File: plugin.video.youtube/resources/lib/youtube_plugin/youtube/provider.py [hunk 7]
```diff
@@ -566,45 +673,204 @@ def _on_yt_get_watch_later_id(self, context, re_match):
...
+                    not xbmc.getInfoLabel('Container.FolderPath').startswith(context.create_uri(['kodion', 'search', 'input'])):
+                # don't refresh search input it causes request for new input, (Container.Update in abstract_provider /kodion/search/input/
+                # would resolve this but doesn't work with Remotes(Yatse))
+                context.get_ui().refresh_container()
         else:
             context.log_warning('Missing video ID for post play event')
         return True
 
+    @kodion.RegisterProviderPath('^/users/(?P<action>[^/]+)/$')
+    def _on_users(self, context, re_match):
```
Comment (enen92) [inline plugin.video.youtube/resources/lib/youtube_plugin/youtube/provider.py]: has this anything to do with supporting several profiles in the addon? Why not simply rely on Kodi's default profile feature? Or is it totally broken? :)

**Comment by anxdpanic** [comment]
Thanks for the review. No worries on the time, was expecting a few day. I will try to respond and address the issues. 

As for the ipstack code, a 'My Location' folder was added that uses latitude, longitude(parameters in youtube api call) to provide geo based results and nothing more.  During the setup wizard, the user is asked "Perform geolocation? Used for My Location"  

If this is not okay, is manual input of this information allowed?

File: plugin.video.youtube/resources/lib/youtube_plugin/youtube/provider.py (hunk 7 above)
Comment (anxdpanic) [reply plugin.video.youtube/resources/lib/youtube_plugin/youtube/provider.py]: Kodi profiles work, this does add multi-account per profile. Been requested a few times by users, and something I use personally.

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py (hunk 1 above)
Comment (anxdpanic) [reply plugin.video.youtube/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py]: icon wasn't added to setArt until Kodi 16

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/service.py (hunk 3 above)
Comment (anxdpanic) [reply plugin.video.youtube/resources/lib/youtube_plugin/kodion/service.py]: added in v14 so should be fine in 15.

**Comment by enen92** [comment]
@anxdpanic thanks for understanding. The problem here is that relying on an API key will also enable api usage stats and along with it...user tracking. I was pointed by @Lunatixz that http://ip-api.com/json could be an alternative (used in plutotv). Would this work for you?

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/utils/function_cache.py (hunk 4 above)
Comment (anxdpanic) [reply plugin.video.youtube/resources/lib/youtube_plugin/kodion/utils/function_cache.py]: It's to override the method from the Storage class. Using database size for function cache and don't want to use item count as well. I'll add a comment to this method while I'm doing the fix up.

File: plugin.video.youtube/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py (hunk 2 above)
Comment (anxdpanic) [reply plugin.video.youtube/resources/lib/youtube_plugin/kodion/impl/xbmc/xbmc_items.py]: Doesn't appear to be needed now, was initially added quite a while ago. One of those times I wish I had added comments. Will be removed in the fix up.

**Comment by anxdpanic** [comment]
@enen92 
ip-api looks like it will do the trick. Will get that changed.

While I'm making these changes, is there anything that needs to be done to identify py3 compatibility?

**Comment by enen92** [comment]
When Leia is final a new field will be available in `addon.xml` to identify the ability of the addon to support both py2 and py3, but as of now it is not yet available. Support for both will be required when submitting to the Leia repository. For M*** py3 is required (py2 no longer used).

You can try to use the addon-check project (https://github.com/xbmc/addon-check) locally against your addon by setting `--branch=leia`. It includes support for `2to3`, a python module that warns about deprecated python usage and suggestions for py3 compatibility. We are only enforcing this check for Krypton and Leia branches of `repo-plugins` and `repo-scripts` IIRC.

**Comment by anxdpanic** [comment]
Changes made and using ip-api now.

**Comment by anxdpanic** [comment]
Thanks for taking the time to review, squashed and updated titles

**Comment by enen92** [comment]
CI failure is already reported in addon-check project and should be addressed soon (https://github.com/xbmc/addon-check/issues/110). Merging.

**Pull request merged**